res/controllers/Traktor-Kontrol-S4-MK3-hid-scripts.js

```diff
@@ -0,0 +1,884 @@
+        this.error("wheel value " + value + " is out of range");
+        return;
+    }
+
+    var tickDelta = this.updateTimeDeltaForWheel(group);
+    var wheelDelta = this.calculateWheelDelta(
+        this.prevWheelValue[group], value);
+
+    var speed = 0;
+    if (tickDelta > 0) {
+        speed = wheelDelta / (tickDelta / 10000.0);
+    }
+    speed *= this.wheelSpeedMultiplier;
+
+    mixxxGroup = "[" + this.translateGroupNames("${Channel}", group) + "]";
+    this.setValue(mixxxGroup, "jog", speed);
+}
+
+// This function is called by the HID packet parser when it has
+// recognised a particular message based on the bits provided to it.
+TraktorS4MK3_alex.wheelsCallback = function(packet, data) {
+    var s4 = TraktorS4MK3_alex
+    for (var name in data) {
+        if (data.hasOwnProperty(name)) {
+            var field = data[name]
+            if (field.id === "message3.clock1") {
+                s4.currWheelTick = field.value
+                s4.currTime = new Date().getTime();
+            }
+            if (field.name == "wheel_long") {
+                s4.updateWheelSpeed(field.group, field.value);
+            }
+        }
+    }
+}
+
+// This function is called by the Mixxx host before it starts sending
+// data packets from the wire to this code.
+TraktorS4MK3_alex.init = function(id) {
+    this.info("initialising...");
+    this.registerS4ToMixxxMapping();
+    this.registerSpecialInputHandlers();
+    this.registerInputPackets()
+    // this.registerOutputPackets()
+    this.info("done init");
+}
+
+// This function is called by the Mixxx host when it is terminating or
+// when it is about to unload this mapping code.
+TraktorS4MK3_alex.shutdown = function() {
+    var packet_lengths = [53, 63, 61];
+    for (i = 0; i < packet_lengths.length; i++) {
+        var packet_length = packet_lengths[i];
+        var data = Object();
+        data.length = packet_length;
+        data[0] = 0x80 + i;
+        for (j = 1; j < packet_length; j++) {
+            data[j] = 0;
+        }
+        // Keep USB light on though.
+        if (i === 0) {
+            data[0x2A] = 0x7F;
+        }
+        controller.send(data, packet_length, 0);
+    }
+}
+
+// This function is called by the Mixxx host when it has received new
+// data from the USB HID interface.
+TraktorS4MK3_alex.incomingData = function(data, length) {
+    // Packets of 23 bytes are message 0x01 and can be handed off immediately.
+    if (length === 23) {
+        this.controller.parsePacket(data, length);
+        return;
+    }
+
+    // Windows seems to get the packet of length 79, so parse as one:
+    if (length == 79) {
+        this.controller.parsePacket(data, data.length);
+        return;
+    }
+
+    // Packets of 64 bytes and 15 bytes are partials. We have to save
+    // the 64 byte portion and then append the 15 bytes when we get it.
+    if (length === 64) {
+        this.partial_packet = data; return;
+    }
+
+    if (length === 15) {
```

Comment:
I will work on a PR to use the hidraw backend of hidapi on Linux so this isn't needed.